cou/steps/plan.py

```diff
@@ -334,26 +358,59 @@ async def generate_plan(analysis_result: Analysis, args: CLIargs) -> UpgradePlan
                 coro=backup(analysis_result.model),
             )
         )
+    return plan
 
-    control_plane_principal_upgrade_plan = await create_upgrade_group(
-        apps=analysis_result.apps_control_plane,
-        description="Control Plane principal(s) upgrade plan",
-        target=target,
-        force=args.force,
-        filter_function=lambda app: not isinstance(app, SubordinateBaseClass),
-    )
-    plan.add_step(control_plane_principal_upgrade_plan)
-
-    control_plane_subordinate_upgrade_plan = await create_upgrade_group(
-        apps=analysis_result.apps_control_plane,
-        description="Control Plane subordinate(s) upgrade plan",
-        target=target,
-        force=args.force,
-        filter_function=lambda app: isinstance(app, SubordinateBaseClass),
-    )
-    plan.add_step(control_plane_subordinate_upgrade_plan)
 
-    return plan
+async def generate_control_plane_plan(
+    analysis_result: Analysis, args: CLIargs, target: OpenStackRelease
+) -> tuple[UpgradePlan, UpgradePlan]:
+    """_summary_.
+
+    :param analysis_result: _description_
+    :type analysis_result: Analysis
+    :param args: _description_
+    :type args: CLIargs
+    :param target: _description_
+    :type target: OpenStackRelease
+    :return: _description_
+    :rtype: tuple[UpgradePlan, UpgradePlan]
```

Comment:
non-blocking: for consistency, I think all the `generate_*` functions should have the same rtype. Instead of a tuple you could return a single plan (with principal and subordinate as substeps)

Reply:
I agree with the consistency and now they are returning `list[UpgradePlan]`. I think returning a simple plan with principal and subordinates as substeps could change the indentation of the plan and break some unit tests.

Reply:
yeah that's the problem with having unit tests that use text representation as a substitute for internal structure... :/

Unit tests should not dictate how we write the rest of the code, btw